src/renderers/dom/fiber/ReactDOMFiberComponent.js

```diff
@@ -69,6 +70,66 @@ if (__DEV__) {
     validateInputPropertes(type, props);
     validateUnknownPropertes(type, props);
   };
+
+  var warnForTextDifference = function(serverText: string, clientText: string) {
+    if (didWarnInvalidHydration) {
+      return;
+    }
+    didWarnInvalidHydration = true;
+    warning(
+      false,
+      'Text content did not match. Server: "%s" Client: "%s"',
+      serverText,
+      clientText,
+    );
+  };
+
+  var warnForPropDifference = function(
+    propName: string,
+    serverValue: mixed,
+    clientValue: mixed,
+  ) {
+    if (didWarnInvalidHydration) {
+      return;
+    }
+    didWarnInvalidHydration = true;
+    warning(
+      false,
+      'Prop `%s` did not match. Server: %s Client: %s',
+      propName,
+      JSON.stringify(serverValue),
+      JSON.stringify(clientValue),
+    );
+  };
+
+  var warnForExtraAttributes = function(attributeNames: Set<string>) {
+    if (didWarnInvalidHydration) {
+      return;
+    }
+    didWarnInvalidHydration = true;
+    var names = [];
+    attributeNames.forEach(function(name) {
+      names.push(name);
+    });
```

Comment:
Just ran the benchmarks from that link locally in node `8.1.3` and the iterator wins there by a tiny bit. I feel like microbenchmarks cannot be trusted in js. Perhaps @bmeurer can chime in?